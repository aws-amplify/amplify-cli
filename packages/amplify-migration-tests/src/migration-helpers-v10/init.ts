import { addCircleCITags, getCLIPath, nspawn as spawn } from '@aws-amplify/amplify-e2e-core';
import { EOL } from 'os';

const defaultSettings = {
  name: EOL,
  // eslint-disable-next-line spellcheck/spell-checker
  envName: 'integtest',
  editor: EOL,
  appType: EOL,
  framework: EOL,
  srcDir: EOL,
  distDir: EOL,
  buildCmd: EOL,
  startCmd: EOL,
  useProfile: EOL,
  profileName: EOL,
  region: process.env.CLI_REGION,
  local: false,
  disableAmplifyAppCreation: true,
  disableCIDetection: false,
  providerConfig: undefined,
  permissionsBoundaryArn: undefined,
};

export function initJSProjectWithProfileV10(cwd: string, settings?: Partial<typeof defaultSettings>): Promise<void> {
  const mergedSettings = { ...defaultSettings, ...settings };
  let env;

  if (mergedSettings.disableAmplifyAppCreation === true) {
    env = {
      CLI_DEV_INTERNAL_DISABLE_AMPLIFY_APP_CREATION: '1',
    };
  }

  addCircleCITags(cwd);

  const cliArgs = ['init'];
  const providerConfigSpecified = !!mergedSettings.providerConfig && typeof mergedSettings.providerConfig === 'object';
  if (providerConfigSpecified) {
    cliArgs.push('--providers', JSON.stringify(mergedSettings.providerConfig));
  }

  if (mergedSettings.permissionsBoundaryArn) {
    cliArgs.push('--permissions-boundary', mergedSettings.permissionsBoundaryArn);
  }

  if (mergedSettings?.name?.length > 20)
    console.warn('Project names should not be longer than 20 characters. This may cause tests to break.');

  return new Promise((resolve, reject) => {
    const chain = spawn(getCLIPath(), cliArgs, {
      cwd,
      stripColors: true,
      env,
      disableCIDetection: mergedSettings.disableCIDetection,
    })
      .wait('Enter a name for the project')
      .sendLine(mergedSettings.name)
      .wait('Initialize the project with the above configuration?')
      .sendConfirmNo()
      .wait('Enter a name for the environment')
      .sendLine(mergedSettings.envName)
      .wait('Choose your default editor:')
      .sendLine(mergedSettings.editor)
      .wait("Choose the type of app that you're building")
      .sendLine(mergedSettings.appType)
      .wait('What javascript framework are you using')
      .sendLine(mergedSettings.framework)
      .wait('Source Directory Path:')
      .sendLine(mergedSettings.srcDir)
      .wait('Distribution Directory Path:')
      .sendLine(mergedSettings.distDir)
      .wait('Build Command:')
      .sendLine(mergedSettings.buildCmd)
      .wait('Start Command:')
      .sendCarriageReturn();

    if (!providerConfigSpecified) {
      chain
        .wait('Using default provider  awscloudformation')
        .wait('Select the authentication method you want to use:')
        .sendCarriageReturn()
        .wait('Please choose the profile you want to use')
        .sendLine(mergedSettings.profileName);
    }
    chain
      .wait('Help improve Amplify CLI by sharing non sensitive configurations on failures')
      .sendYes()
      .wait(/Try "amplify add api" to create a backend API and then "amplify (push|publish)" to deploy everything/)
      .run((err: Error) => {
        if (err) {
          reject(err);
        } else {
          resolve();
        }
      });
  });
}
