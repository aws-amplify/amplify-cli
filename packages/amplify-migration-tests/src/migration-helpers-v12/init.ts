import { getCLIPath, nspawn as spawn } from '@aws-amplify/amplify-e2e-core';
import { EOL } from 'os';

const defaultSettings = {
  name: EOL,
  // eslint-disable-next-line spellcheck/spell-checker
  envName: 'integtest',
  editor: EOL,
  appType: EOL,
  framework: EOL,
  srcDir: EOL,
  distDir: EOL,
  buildCmd: EOL,
  startCmd: EOL,
  useProfile: EOL,
  profileName: EOL,
  region: process.env.CLI_REGION,
  local: false,
  disableAmplifyAppCreation: true,
  disableCIDetection: false,
  providerConfig: undefined,
  permissionsBoundaryArn: undefined,
};

export function initJSProjectWithProfileV12(cwd: string, settings?: Partial<typeof defaultSettings>): Promise<void> {
  const s = { ...defaultSettings, ...settings };
  let env;

  if (s.disableAmplifyAppCreation === true) {
    env = {
      CLI_DEV_INTERNAL_DISABLE_AMPLIFY_APP_CREATION: '1',
    };
  }

  const cliArgs = ['init'];
  const providerConfigSpecified = !!s.providerConfig && typeof s.providerConfig === 'object';
  if (providerConfigSpecified) {
    cliArgs.push('--providers', JSON.stringify(s.providerConfig));
  }

  if (s.permissionsBoundaryArn) {
    cliArgs.push('--permissions-boundary', s.permissionsBoundaryArn);
  }

  if (s?.name?.length > 20) console.warn('Project names should not be longer than 20 characters. This may cause tests to break.');

  const chain = spawn(getCLIPath(), cliArgs, {
    cwd,
    stripColors: true,
    env,
    disableCIDetection: s.disableCIDetection,
  })
    .wait('Enter a name for the project')
    .sendLine(s.name)
    .wait('Initialize the project with the above configuration?')
    .sendConfirmNo()
    .wait('Enter a name for the environment')
    .sendLine(s.envName)
    .wait('Choose your default editor:')
    .sendLine(s.editor)
    .wait("Choose the type of app that you're building")
    .sendLine(s.appType)
    .wait('What javascript framework are you using')
    .sendLine(s.framework)
    .wait('Source Directory Path:')
    .sendLine(s.srcDir)
    .wait('Distribution Directory Path:')
    .sendLine(s.distDir)
    .wait('Build Command:')
    .sendLine(s.buildCmd)
    .wait('Start Command:')
    .sendCarriageReturn();

  if (!providerConfigSpecified) {
    chain
      .wait('Using default provider  awscloudformation')
      .wait('Select the authentication method you want to use:')
      .sendCarriageReturn()
      .wait('Please choose the profile you want to use')
      .sendLine(s.profileName);
  }
  return chain
    .wait(/Help improve Amplify CLI by sharing non( |-)sensitive( | project )configurations on failures/)
    .sendYes()
    .wait(/Try "amplify add api" to create a backend API and then "amplify (push|publish)" to deploy everything/)
    .runAsync();
}

export function initIosProjectWithProfileV12(cwd: string, settings: Record<string, unknown>): Promise<void> {
  const s = { ...defaultSettings, ...settings };

  let env;

  if (s.disableAmplifyAppCreation === true) {
    env = {
      CLI_DEV_INTERNAL_DISABLE_AMPLIFY_APP_CREATION: '1',
    };
  }

  return spawn(getCLIPath(), ['init'], {
    cwd,
    stripColors: true,
    env,
  })
    .wait('Do you want to continue with Amplify Gen 1?')
    .sendYes()
    .wait('Why would you like to use Amplify Gen 1?')
    .sendCarriageReturn()
    .wait('Enter a name for the project')
    .sendLine(s.name)
    .wait('Initialize the project with the above configuration?')
    .sendConfirmNo()
    .wait('Enter a name for the environment')
    .sendLine(s.envName)
    .wait('Choose your default editor:')
    .sendLine(s.editor)
    .wait("Choose the type of app that you're building")
    .sendKeyDown(3)
    .sendCarriageReturn()
    .wait('Select the authentication method you want to use:')
    .sendCarriageReturn()
    .wait('Please choose the profile you want to use')
    .sendLine(s.profileName)
    .wait(/Help improve Amplify CLI by sharing non( |-)sensitive( | project )configurations on failures/)
    .sendYes()
    .wait(/Try "amplify add api" to create a backend API and then "amplify (push|publish)" to deploy everything/)
    .runAsync();
}

export function initAndroidProjectWithProfileV12(cwd: string, settings: Partial<typeof defaultSettings>): Promise<void> {
  const s = { ...defaultSettings, ...settings };

  let env;

  if (s.disableAmplifyAppCreation) {
    env = {
      CLI_DEV_INTERNAL_DISABLE_AMPLIFY_APP_CREATION: '1',
    };
  }

  return spawn(getCLIPath(), ['init'], {
    cwd,
    stripColors: true,
    env,
  })
    .wait('Enter a name for the project')
    .sendLine(s.name)
    .wait('Initialize the project with the above configuration?')
    .sendConfirmNo()
    .wait('Enter a name for the environment')
    .sendLine(s.envName)
    .wait('Choose your default editor:')
    .sendLine(s.editor)
    .wait("Choose the type of app that you're building")
    .sendKeyDown(1)
    .sendCarriageReturn()
    .wait('Where is your Res directory')
    .sendCarriageReturn()
    .wait('Select the authentication method you want to use:')
    .sendCarriageReturn()
    .wait('Please choose the profile you want to use')
    .sendLine(s.profileName)
    .wait(/Help improve Amplify CLI by sharing non( |-)sensitive( | project )configurations on failures/)
    .sendYes()
    .wait(/Try "amplify add api" to create a backend API and then "amplify (push|publish)" to deploy everything/)
    .runAsync();
}
