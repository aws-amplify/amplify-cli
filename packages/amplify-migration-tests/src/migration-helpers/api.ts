import {
  AddApiOptions,
  defaultOptions,
  getCLIPath,
  getSchemaPath,
  KEY_DOWN_ARROW,
  nspawn as spawn,
  selectRuntime,
  selectTemplate,
} from 'amplify-e2e-core';
import { assign } from 'lodash';

/**
 * Old Dx prior to this api workflow change https://github.com/aws-amplify/amplify-cli/pull/8153
 */
export function addApiWithoutSchemaOldDx(cwd: string, opts: Partial<AddApiOptions> = {}) {
  const options = assign(defaultOptions, opts);
  return new Promise<void>((resolve, reject) => {
    spawn(getCLIPath(), ['add', 'api'], { cwd, stripColors: true })
      .wait('Please select from one of the below mentioned services:')
      .sendCarriageReturn()
      .wait('Provide API name:')
      .sendLine(options.apiName)
      .wait(/.*Choose the default authorization type for the API.*/)
      .sendCarriageReturn()
      .wait(/.*Enter a description for the API key.*/)
      .sendCarriageReturn()
      .wait(/.*After how many days from now the API key should expire.*/)
      .sendCarriageReturn()
      .wait(/.*Do you want to configure advanced settings for the GraphQL API.*/)
      .sendCarriageReturn()
      .wait('Do you have an annotated GraphQL schema?')
      .sendConfirmNo()
      .wait('Choose a schema template:')
      .sendCarriageReturn()
      .wait('Do you want to edit the schema now?')
      .sendConfirmNo()
      .wait(
        '"amplify publish" will build all your local backend and frontend resources (if you have hosting category added) and provision it in the cloud',
      )
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}

export function addApiWithSchemaAndConflictDetectionOldDx(cwd: string, schemaFile: string) {
  const schemaPath = getSchemaPath(schemaFile);
  return new Promise<void>((resolve, reject) => {
    spawn(getCLIPath(), ['add', 'api'], { cwd, stripColors: true })
      .wait('Please select from one of the below mentioned services:')
      .sendCarriageReturn()
      .wait('Provide API name:')
      .sendCarriageReturn()
      .wait(/.*Choose the default authorization type for the API.*/)
      .sendCarriageReturn()
      .wait(/.*Enter a description for the API key.*/)
      .sendCarriageReturn()
      .wait(/.*After how many days from now the API key should expire.*/)
      .sendCarriageReturn()
      .wait(/.*Do you want to configure advanced settings for the GraphQL API.*/)
      .sendLine(KEY_DOWN_ARROW) // Down
      .wait(/.*Configure additional auth types.*/)
      .sendConfirmNo()
      .wait(/.*Enable conflict detection.*/)
      .sendConfirmYes()
      .wait(/.*Select the default resolution strategy.*/)
      .sendCarriageReturn()
      .wait(/.*Do you have an annotated GraphQL schema.*/)
      .sendConfirmYes()
      .wait('Provide your schema file path:')
      .sendLine(schemaPath)
      .wait(
        '"amplify publish" will build all your local backend and frontend resources (if you have hosting category added) and provision it in the cloud',
      )
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}

export function addRestApiOldDx(cwd: string, settings: any) {
  const isFirstRestApi = settings.isFirstRestApi ?? true;
  let chain = spawn(getCLIPath(), ['add', 'api'], { cwd, stripColors: true })
    .wait('Select from one of the below mentioned services')
    .sendKeyDown()
    .sendCarriageReturn(); // REST

  if (!isFirstRestApi) {
    chain.wait('Would you like to add a new path to an existing REST API');

    if (settings.path) {
      chain
        .sendYes()
        .wait('Select the REST API you want to update')
        .sendCarriageReturn() // Select the first REST API
        .wait('What would you like to do?')
        .sendCarriageReturn() // Add another path
        .wait('Provide a path')
        .sendLine(settings.path)
        .wait('Choose a lambda source')
        .sendKeyDown()
        .sendCarriageReturn() // Existing lambda
        .wait('Choose the Lambda function to invoke by this path')
        .sendCarriageReturn() // Pick first one
        .wait('Restrict API access')
        .sendConfirmNo() // Do not restrict access
        .wait('Do you want to add another path')
        .sendConfirmNo() // Do not add another path
        .sendEof();

      return chain.runAsync();
    } else {
      chain.sendConfirmNo();
    }
  }

  chain.wait('Provide a friendly name for your resource to be used as a label for this category in the project');
  if (settings.apiName) {
    chain.sendLine(settings.apiName);
  } else {
    chain.sendCarriageReturn();
  }
  chain.wait('Provide a path').sendCarriageReturn().wait('Choose a lambda source');

  if (settings.existingLambda) {
    chain
      .sendKeyDown()
      .sendCarriageReturn() // Existing lambda
      .wait('Choose the Lambda function to invoke by this path'); // Expect only 1 Lambda is present
  } else {
    chain
      .sendCarriageReturn() // Create new Lambda function
      .wait('Provide an AWS Lambda function name')
      .sendCarriageReturn();

    selectRuntime(chain, 'nodejs');

    const templateName = settings.isCrud
      ? 'CRUD function for DynamoDB (Integration with API Gateway)'
      : 'Serverless ExpressJS function (Integration with API Gateway)';
    selectTemplate(chain, templateName, 'nodejs');

    if (settings.isCrud) {
      chain
        .wait('Choose a DynamoDB data source option')
        .sendCarriageReturn() // Use DDB table configured in current project
        .wait('Choose from one of the already configured DynamoDB tables')
        .sendCarriageReturn(); // Use first one in the list
    }

    chain
      .wait('Do you want to configure advanced settings?')
      .sendConfirmNo()
      .wait('Do you want to edit the local lambda function now')
      .sendConfirmNo();
  }

  chain.wait('Restrict API access');
  if (settings.restrictAccess) {
    chain.sendConfirmYes();

    if (settings.hasUserPoolGroups) {
      chain.wait('Restrict access by').sendCarriageReturn(); // Auth/Guest Users
    }

    chain.wait('Who should have access');

    if (settings.allowGuestUsers) {
      chain
        .sendKeyDown()
        .sendCarriageReturn() // Authenticated and Guest users
        .wait('What permissions do you want to grant to Authenticated users')
        .send('a') // CRUD permissions for authenticated users
        .sendCarriageReturn()
        .wait('What permissions do you want to grant to Guest users')
        .send('a') // CRUD permissions for guest users
        .sendCarriageReturn();
    } else {
      chain
        .sendCarriageReturn() // Authenticated users only
        .wait('What permissions do you want to grant to Authenticated users')
        .send('a') // CRUD permissions
        .sendCarriageReturn();
    }
  } else {
    chain.sendConfirmNo(); // Do not restrict access
  }

  chain.wait('Do you want to add another path').sendConfirmNo().sendEof();

  return chain.runAsync();
}
