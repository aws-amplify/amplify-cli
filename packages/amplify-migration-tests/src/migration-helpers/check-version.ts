import { getCLIPath, nspawn as spawn } from '@aws-amplify/amplify-e2e-core';

export function versionCheck(cwd: string, testingWithLatestCodebase = false, version: { v?: string } = {}): Promise<void> {
  return new Promise((resolve, reject) => {
    spawn(getCLIPath(testingWithLatestCodebase), ['-v'], { cwd, stripColors: true })
      .wait(/\d+\.\d+\.\d+/, (v) => (version.v = v.trim()))
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}

/**
 * Validates from and to versions for migration tests.
 */
export const validateVersionsForMigrationTest = async (): Promise<void> => {
  const migrateFromVersion = { v: 'uninitialized' };
  const migrateToVersion = { v: 'uninitialized' };
  await versionCheck(process.cwd(), false, migrateFromVersion);
  await versionCheck(process.cwd(), true, migrateToVersion);
  console.log(`Test migration from: ${migrateFromVersion.v} to ${migrateToVersion.v}`);
  expect(migrateFromVersion.v).not.toEqual(migrateToVersion.v);
  expect(allowedVersionsToMigrateFrom).toContain(migrateFromVersion.v);
};

/**
 * This list is used to check migration tests with the following changes. (excludes layer migration tests)
 *
 * api add/update flow: https://github.com/aws-amplify/amplify-cli/pull/8287
 *
 * ext migrate flow: https://github.com/aws-amplify/amplify-cli/pull/8806
 */
export const allowedVersionsToMigrateFrom = ['8.2.0', '10.5.1', '12.0.3'];
