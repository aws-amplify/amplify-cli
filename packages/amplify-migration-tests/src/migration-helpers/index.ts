export * from './init';
export * from './api';
export * from './check-version';
export * from './legacy-lambda-layer';
