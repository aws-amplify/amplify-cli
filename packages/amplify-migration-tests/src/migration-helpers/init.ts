import { addCircleCITags, getCLIPath, nspawn as spawn } from 'amplify-e2e-core';
import { EOL } from 'os';

const defaultSettings = {
  name: EOL,
  envName: 'integtest',
  editor: EOL,
  appType: EOL,
  framework: EOL,
  srcDir: EOL,
  distDir: EOL,
  buildCmd: EOL,
  startCmd: EOL,
  useProfile: EOL,
  profileName: EOL,
  region: process.env.CLI_REGION,
  local: false,
  disableAmplifyAppCreation: true,
};

export function initJSProjectWithProfileOldDX(cwd: string, settings: Object, testingWithLatestCodebase = false): Promise<void> {
  const s = { ...defaultSettings, ...settings };
  let env;

  if (s.disableAmplifyAppCreation === true) {
    env = {
      CLI_DEV_INTERNAL_DISABLE_AMPLIFY_APP_CREATION: '1',
    };
  }

  addCircleCITags(cwd);

  return new Promise((resolve, reject) => {
    spawn(getCLIPath(testingWithLatestCodebase), ['init'], { cwd, stripColors: true, env })
      .wait('Enter a name for the project')
      .sendLine(s.name)
      .wait('Enter a name for the environment')
      .sendLine(s.envName)
      .wait('Choose your default editor:')
      .sendLine(s.editor)
      .wait("Choose the type of app that you're building")
      .sendLine(s.appType)
      .wait('What javascript framework are you using')
      .sendLine(s.framework)
      .wait('Source Directory Path:')
      .sendLine(s.srcDir)
      .wait('Distribution Directory Path:')
      .sendLine(s.distDir)
      .wait('Build Command:')
      .sendLine(s.buildCmd)
      .wait('Start Command:')
      .sendCarriageReturn()
      .wait('Using default provider  awscloudformation')
      .wait(/(Select the authentication method you want to use|Do you want to use an AWS profile)/)
      .sendCarriageReturn()
      .wait('Please choose the profile you want to use')
      .sendLine(s.profileName)
      .wait(/Try "amplify add api" to create a backend API and then "amplify (push|publish)" to deploy everything/)
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}

export function initJSProjectWithProfile(cwd: string, settings: Object, testingWithLatestCodebase = false): Promise<void> {
  const s = { ...defaultSettings, ...settings };
  let env;

  if (s.disableAmplifyAppCreation === true) {
    env = {
      CLI_DEV_INTERNAL_DISABLE_AMPLIFY_APP_CREATION: '1',
    };
  }

  addCircleCITags(cwd);

  return new Promise((resolve, reject) => {
    spawn(getCLIPath(testingWithLatestCodebase), ['init'], { cwd, stripColors: true, env })
      .wait('Enter a name for the project')
      .sendLine(s.name)
      .wait('Initialize the project with the above configuration?')
      .sendConfirmNo()
      .wait('Enter a name for the environment')
      .sendLine(s.envName)
      .wait('Choose your default editor:')
      .sendLine(s.editor)
      .wait("Choose the type of app that you're building")
      .sendLine(s.appType)
      .wait('What javascript framework are you using')
      .sendLine(s.framework)
      .wait('Source Directory Path:')
      .sendLine(s.srcDir)
      .wait('Distribution Directory Path:')
      .sendLine(s.distDir)
      .wait('Build Command:')
      .sendLine(s.buildCmd)
      .wait('Start Command:')
      .sendCarriageReturn()
      .wait('Using default provider  awscloudformation')
      .wait(/(Select the authentication method you want to use|Do you want to use an AWS profile)/)
      .sendCarriageReturn()
      .wait('Please choose the profile you want to use')
      .sendLine(s.profileName)
      .wait(/Try "amplify add api" to create a backend API and then "amplify (push|publish)" to deploy everything/)
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}
