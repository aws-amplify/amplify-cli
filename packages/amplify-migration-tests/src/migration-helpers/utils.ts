import {
  amplifyPull,
  amplifyPushForce,
  amplifyPushWithoutCodegen,
  cliInputsExists,
  createNewProjectDir,
  deleteProjectDir,
  getAppId,
  getBackendConfig,
  getCLIInputs,
  getCloudFormationTemplate,
  getParameters,
  parametersExists,
} from '@aws-amplify/amplify-e2e-core';
import * as cfnDiff from '@aws-cdk/cloudformation-diff';
import { Writable } from 'stream';
import { AmplifyCategories } from '@aws-amplify/amplify-cli-core';
import strip from 'strip-ansi';
import { v4 as uuid } from 'uuid';

/**
 * generates a random string
 */
export const getShortId = (): string => {
  const [shortId] = uuid().split('-');

  return shortId;
};

/**
 * Given the parameter objects of each project for the provided category & resource key,
 * you can modify the parameter objects before the diff is performed.
 *
 * You can use conditional logic to delete attributes on the parameter objects before they are diffed, if
 * you want to exclude those attributes from the comparison. Make sure to return the modified objects
 * that you want to be diffed.
 */
export type ExcludeFromParameterDiff = (
  currentCategory: string,
  currentResourceKey: string,
  parameters: {
    project1: Record<string, unknown>;
    project2: Record<string, unknown>;
  },
) => { project1: Record<string, unknown>; project2: Record<string, unknown> };

/**
 * Asserts that parameters between two project directories didn't drift.
 */
export const assertNoParameterChangesBetweenProjects = (
  projectRoot1: string,
  projectRoot2: string,
  options?: {
    excludeFromParameterDiff?: ExcludeFromParameterDiff;
  },
): void => {
  const backendConfig1 = getBackendConfig(projectRoot1);
  const backendConfig2 = getBackendConfig(projectRoot2);
  expect(backendConfig2).toMatchObject(backendConfig1);
  for (const categoryKey of Object.keys(backendConfig1)) {
    const category = backendConfig1[categoryKey];
    if (Object.values(AmplifyCategories).includes(categoryKey)) {
      for (const resourceKey of Object.keys(category)) {
        if (cliInputsExists(projectRoot1, categoryKey, resourceKey)) {
          const cliInputs1 = getCLIInputs(projectRoot1, categoryKey, resourceKey);
          const cliInputs2 = getCLIInputs(projectRoot2, categoryKey, resourceKey);
          expect(cliInputs1).toEqual(cliInputs2);
        }
        if (parametersExists(projectRoot1, categoryKey, resourceKey)) {
          let parameters1 = getParameters(projectRoot1, categoryKey, resourceKey);
          let parameters2 = getParameters(projectRoot2, categoryKey, resourceKey);
          if (options && options.excludeFromParameterDiff) {
            const afterExclusions = options.excludeFromParameterDiff(categoryKey, resourceKey, {
              project1: parameters1,
              project2: parameters2,
            });
            parameters1 = afterExclusions.project1;
            parameters2 = afterExclusions.project2;
          }
          expect(parameters1).toEqual(parameters2);
        }
      }
    }
  }
};

class InMemoryWritable extends Writable {
  private _payload = '';

  _write(chunk: unknown, __encoding: BufferEncoding, callback: (error?: Error | null) => void): void {
    if (chunk) {
      this._payload += chunk.toString();
    }
    callback();
  }

  toString(): string {
    return this._payload;
  }
}

/**
 * Given the CFN templates of each project for the provided category & resource key,
 * you can modify the CFN templates objects before the diff is performed.
 *
 * You can use conditional logic to delete attributes on the CFN object before they are diffed, if
 * you want to exclude those attributes from the comparison. Make sure to return the modified objects
 * that you want to be diffed.
 */
export type ExcludeFromCFNDiff = (
  currentCategory: string,
  currentResourceKey: string,
  cfnTemplates: {
    project1: Record<string, unknown>;
    project2: Record<string, unknown>;
  },
) => { project1: Record<string, unknown>; project2: Record<string, unknown> };

/**
 * Collects all differences between cloud formation templates into a single string.
 */
export const collectCloudformationDiffBetweenProjects = (
  projectRoot1: string,
  projectRoot2: string,
  excludeFn?: ExcludeFromCFNDiff,
): string => {
  const backendConfig1 = getBackendConfig(projectRoot1);
  const backendConfig2 = getBackendConfig(projectRoot2);
  expect(backendConfig2).toMatchObject(backendConfig1);
  const stream = new InMemoryWritable();
  for (const categoryKey of Object.keys(backendConfig1)) {
    const category = backendConfig1[categoryKey];
    for (const resourceKey of Object.keys(category)) {
      let template1 = getCloudFormationTemplate(projectRoot1, categoryKey, resourceKey);
      let template2 = getCloudFormationTemplate(projectRoot2, categoryKey, resourceKey);

      // Description does not matter much and it can contain os/runtime specific words.
      delete template1.Description;
      delete template2.Description;

      if (excludeFn) {
        const afterExclusions = excludeFn(categoryKey, resourceKey, { project1: template1, project2: template2 });
        template1 = afterExclusions.project1;
        template2 = afterExclusions.project2;
      }

      const templateDiff = cfnDiff.diffTemplate(template1, template2);
      if (!templateDiff.isEmpty) {
        cfnDiff.formatDifferences(stream, templateDiff);
      }
    }
  }
  return strip(stream.toString());
};

/**
 * Pulls and pushes project with latest codebase. Validates parameter and cfn drift.
 */
export const pullPushWithLatestCodebaseValidateParameterAndCfnDrift = async (projRoot: string, projName: string): Promise<void> => {
  const appId = getAppId(projRoot);
  expect(appId).toBeDefined();
  const projRoot2 = await createNewProjectDir(`${projName}2`);
  try {
    await amplifyPull(projRoot2, { emptyDir: true, appId }, true);
    assertNoParameterChangesBetweenProjects(projRoot, projRoot2);
    expect(collectCloudformationDiffBetweenProjects(projRoot, projRoot2)).toMatchSnapshot();
    await amplifyPushWithoutCodegen(projRoot2, true);
    assertNoParameterChangesBetweenProjects(projRoot, projRoot2);
    expect(collectCloudformationDiffBetweenProjects(projRoot, projRoot2)).toMatchSnapshot();
  } finally {
    deleteProjectDir(projRoot2);
  }
};

/**
 * Pulls and pushes with force project with latest codebase. Validates parameter and cfn drift.
 */
export const pullPushForceWithLatestCodebaseValidateParameterAndCfnDrift = async (projRoot: string, projRoot2: string): Promise<void> => {
  const appId = getAppId(projRoot);
  expect(appId).toBeDefined();
  try {
    await amplifyPull(projRoot2, { emptyDir: true, appId }, true);
    assertNoParameterChangesBetweenProjects(projRoot, projRoot2);
    expect(collectCloudformationDiffBetweenProjects(projRoot, projRoot2)).toMatchSnapshot();
    await amplifyPushForce(projRoot2, true);
    assertNoParameterChangesBetweenProjects(projRoot, projRoot2);
    expect(collectCloudformationDiffBetweenProjects(projRoot, projRoot2)).toMatchSnapshot();
  } finally {
    deleteProjectDir(projRoot2);
  }
};
