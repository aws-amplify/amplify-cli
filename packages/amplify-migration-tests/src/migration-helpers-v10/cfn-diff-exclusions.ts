import { ExcludeFromCFNDiff } from '../migration-helpers/utils';

/**
 * Due to a known limitation of APIGateway, we do not want the AWS::ApiGateway::Deployment resource
 * from being compared in the CFN diff because it is regenerated whenever an amplify app is pulled down.
 * This would produce a false positive when checking for differences in the CFN templates of project1 & project2.
 * https://github.com/aws/aws-cdk/issues/8646#issuecomment-647561856
 *
 * AWS::ApiGateway::GatewayResponse is also excluded
 */
export const cfnDiffExclusions: ExcludeFromCFNDiff = (
  currentCategory: string,
  currentResourceKey: string,
  cfnTemplates: {
    project1: Record<string, unknown>;
    project2: Record<string, unknown>;
  },
) => {
  const excludeAPIGateWayDeploymentResource = (cfnTemplate: Record<string, unknown>): void => {
    const resources = cfnTemplate.Resources ?? {};
    const resourceKeys = Object.keys(resources);
    for (const key of resourceKeys) {
      const resource = resources[key];
      if (resource.Type === 'AWS::ApiGateway::Deployment' || resource.Type === 'AWS::ApiGateway::GatewayResponse') {
        delete resources[key];
      }
      if (resource.Type === 'AWS::AppSync::ApiKey' && resource.Properties) {
        delete resource.Properties.Expires;
      }
    }
  };
  if (currentCategory === 'api') {
    excludeAPIGateWayDeploymentResource(cfnTemplates.project1);
    excludeAPIGateWayDeploymentResource(cfnTemplates.project2);
  }
  return { project1: cfnTemplates.project1, project2: cfnTemplates.project2 };
};
