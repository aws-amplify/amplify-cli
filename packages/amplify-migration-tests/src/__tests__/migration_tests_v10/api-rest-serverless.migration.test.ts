import {
  addRestApi,
  amplifyPull,
  amplifyPushAuth,
  amplifyPushUpdateLegacy,
  createNewProjectDir,
  deleteProject,
  deleteProjectDir,
  getAppId,
  getProjectMeta,
  validateRestApiMeta,
} from '@aws-amplify/amplify-e2e-core';
import { cfnDiffExclusions } from '../../migration-helpers-v10/cfn-diff-exclusions';
import { initJSProjectWithProfileV10 } from '../../migration-helpers-v10/init';
import { assertNoParameterChangesBetweenProjects, collectCloudformationDiffBetweenProjects } from '../../migration-helpers/utils';

describe('api serverless migration tests', () => {
  let projRoot: string;
  let projectName: string;

  afterEach(async () => {
    await deleteProject(projRoot, undefined, true);
    deleteProjectDir(projRoot);
  });

  it('...adds serverless REST api with v10 and pulls without drift in latest version', async () => {
    projectName = 'restDDB';
    projRoot = await createNewProjectDir(projectName);

    await initJSProjectWithProfileV10(projRoot, { name: 'restApiTest', disableAmplifyAppCreation: false });
    await addRestApi(projRoot, { isCrud: false });
    await amplifyPushUpdateLegacy(projRoot);

    const meta = getProjectMeta(projRoot);
    validateRestApiMeta(projRoot, meta);

    // pull down with vlatest
    const appId = getAppId(projRoot);
    expect(appId).toBeDefined();
    const projRoot2 = await createNewProjectDir(`${projectName}2`);
    try {
      await amplifyPull(projRoot2, { emptyDir: true, appId }, true);
      assertNoParameterChangesBetweenProjects(projRoot, projRoot2);
      expect(collectCloudformationDiffBetweenProjects(projRoot, projRoot2, cfnDiffExclusions)).toMatchSnapshot();
      await amplifyPushAuth(projRoot2, true);
      assertNoParameterChangesBetweenProjects(projRoot, projRoot2);
      expect(collectCloudformationDiffBetweenProjects(projRoot, projRoot2, cfnDiffExclusions)).toMatchSnapshot();

      // validate metadata
      const meta2 = getProjectMeta(projRoot2);
      validateRestApiMeta(projRoot2, meta2);
    } finally {
      deleteProjectDir(projRoot2);
    }
  });
});
