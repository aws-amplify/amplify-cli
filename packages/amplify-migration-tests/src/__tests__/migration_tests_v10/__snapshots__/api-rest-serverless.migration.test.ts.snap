// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api serverless migration tests ...adds serverless REST api with v10 and pulls without drift in latest version 1`] = `""`;

exports[`api serverless migration tests ...adds serverless REST api with v10 and pulls without drift in latest version 2`] = `
"Outputs
[+] Output LambdaExecutionRoleArn: {\"Value\":{\"Fn::GetAtt\":[\"LambdaExecutionRole\",\"Arn\"]}}

"
`;
